File: Python Client/GameGRPCClient/GuessNumberGameClient.py
from concurrent import futures
import game_pb2 as guess_number_game_pb2
import game_pb2_grpc as guess_number_game_pb2_grpc
import grpc
import sys

class GuessNumberGameClient:
    def __init__(self, host, port):
        channel = grpc.insecure_channel(f"{host}:{port}")
        self.stub = guess_number_game_pb2_grpc.GuessNumberGameStub(channel)
        self.latch = futures.Future()

    def hasWinner(self, response):
        winner = response.winner
        return winner if winner != 0 else -1

    def start_game(self):
        def handle_response(response):
            print(response.message)
            winner = self.hasWinner(response)
            if winner >= 0:
                print(f"Player {winner} wins!")
                self.latch.set_result(True)

        request_iterator = self.generate_requests()
        response_iterator = self.stub.GuessNumber(request_iterator)

        try:
            for response in response_iterator:
                handle_response(response)
                if self.latch.done():
                    break
        except grpc.RpcError:
            print("The Game Has been Ended")

    def generate_requests(self):
        while True:
            try:
                number = int(input("Enter your guess (between 1 and 1000): "))
                yield guess_number_game_pb2.GuessRequest(number=number)
            except Exception as e:
                print("Error: " + str(e))
                sys.exit(1)

    def shutdown(self):
        pass


if __name__ == "__main__":
    host = "localhost"
    port = 9000

    client = GuessNumberGameClient(host, port)

    try:
        client.start_game()
    except KeyboardInterrupt:
        pass
    finally:
        client.shutdown()

File: GameJavaClient/src/main/java/GuessNumberGameClient.java
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import ma.Game.grpc.GuessNumberGameGrpc;
import ma.Game.grpc.GuessRequest;
import ma.Game.grpc.GuessResponse;

import java.util.Scanner;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class GuessNumberGameClient {
    private final ManagedChannel channel;
    private final GuessNumberGameGrpc.GuessNumberGameStub stub;

    public GuessNumberGameClient(String host, int port) {
        channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
        stub = GuessNumberGameGrpc.newStub(channel);
    }
    private static int hasWinner(GuessResponse response) {
        int winner = response.getWinner();
        return winner != 0 ? winner : -1;
    }


    public void startGame() throws InterruptedException {
        CountDownLatch latch = new CountDownLatch(1);

        StreamObserver<GuessRequest> requestObserver = stub.guessNumber(new StreamObserver<GuessResponse>() {
            @Override
            public void onNext(GuessResponse response) {
                System.out.println(response.getMessage());
                int winner = hasWinner(response);
                if (winner != -1) {
                    System.out.println("Player " + winner + " wins!");
                    latch.countDown();
                }
            }


            @Override
            public void onError(Throwable throwable) {
                System.out.println("Error: " + throwable.getMessage());
                channel.shutdown();
                latch.countDown();
            }

            @Override
            public void onCompleted() {
                System.out.println("Game over!");
                channel.shutdown();
                latch.countDown();
            }
        });

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Enter your guess (between 1 and 1000):");
            int number = scanner.nextInt();

            requestObserver.onNext(GuessRequest.newBuilder().setNumber(number).build());

            if (latch.getCount() == 0) {
                break;
            }
        }

        requestObserver.onCompleted();

        latch.await();
    }

    public void shutdown() throws InterruptedException {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
    }

    public static void main(String[] args) throws InterruptedException {
        String host = "localhost";
        int port = 9000;

        GuessNumberGameClient client = new GuessNumberGameClient(host, port);

        try {
            client.startGame();
        } finally {
            client.shutdown();
        }
    }
}